pkg/mappings.bzl

```diff
@@ -591,3 +611,117 @@ pkg_filegroup = rule(
     },
     provides = [PackageFilegroupInfo],
 )
+
+def _filter_directory_argify_pair(pair):
+    return "{}={}".format(*pair)
+
+def _filter_directory_impl(ctx):
+    out_dir = ctx.actions.declare_directory(ctx.attr.outdir_name or ctx.attr.name)
+
+    if not ctx.file.src.is_directory:
+        fail("Must be a directory (TreeArtifact)", "src")
+
+    args = ctx.actions.args()
+
+    # Flags
+    args.add_all(ctx.attr.excludes, before_each = "--exclude")
+    args.add_all(ctx.attr.renames.items(), before_each = "--rename", map_each = _filter_directory_argify_pair)
+
+    args.add("--prefix", ctx.attr.prefix)
+    args.add("--strip_prefix", ctx.attr.strip_prefix)
+
+    # Adding the directories directly here requires manually specifying the
+    # path.  Bazel will reject simply passing in the File object.
+    args.add(ctx.file.src.path)
+    args.add(out_dir.path)
+
+    ctx.actions.run(
+        executable = ctx.executable._filterer,
+        use_default_shell_env = True,
+        arguments = [args],
+        inputs = [ctx.file.src],
+        outputs = [out_dir],
+    )
+
+    return [DefaultInfo(files = depset([out_dir]))]
+
+filter_directory = rule(
+    doc = """
+    Transform directories (TreeArtifacts) using pkg_filegroup-like semantics.
+
+    Order of operations:
+    
+    - Files are `exclude`d
+    - `strip_prefix` is applied
+    - `renames` is applied, overriding `strip_prefix`
```

Comment:
I don't follow that.   If strip_prefix is first, how can the second operation override it?
What would happen if I had
strip_prefix='foo',
renames='foo/bar'=>'baz'

Would we end up with baz?  Or, because we had stripped first, not matched on foo/bar?

Reply:
`baz`. `strip_prefix` is ignored when there is a matching `rename`.

I'll restructure the documentation a bit.

Reply:
Done.  PTAL.